keras_nlp/layers/mlm_classification_head.py

```diff
@@ -0,0 +1,204 @@
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+class MLMClassificationHead(keras.layers.Layer):
+    """Masked Language Model classification head.
+
+    This layer takes two inputs:
+     - `inputs`: which should be a tensor of encoded tokens with shape
+            `(batch_size, sequence_length, encoding_dim)`.
+     - `mask_positions`: with be a tensor of integer positions to predict
+            with shape `(batch_size, masks_per_sequence)`.
+
+    The token encodings should usually be the last output of an encoder model,
+    and mask positions should be the interger positions you would like to
+    predict for the MLM task.
+
+    The layer will first gather the token encodings at the mask positions. These
+    gathered tokens will be passed through a dense layer the same size as
+    encoding dimension, then transformed to predictions the same size as the
+    input vocabulary.  This layer will produce a single output with shape
+    `(batch_size, masks_per_sequence, vocabulary_size)`, which can be used to
+    compute an MLM loss function.
+
+    Args:
+        vocabulary_size: The total size of the vocabulary for predictions.
+        embedding_weights: Optional. The weights of the word embedding used
+            to transform input token ids. The transpose of this weight matrix
+            will be used to project a token embedding vector to a prediction
+            over all input words, as described in [1].
+        inner_activation: The activation function of inner dense layer.
+        outer_activation: The activation function for the outputs of the layer.
+            Usually either `None` (return logits), or `"softmax"`
+            (return probabilities).
+        layer_norm_epsilon: float, defaults to 1e-5. The epsilon value in layer
+            normalization components.
+        kernel_initializer: string or tf.keras.initializers initializer,
+            defaults to "glorot_uniform". The kernel initializer for
+            the dense and multiheaded attention layers.
+        bias_initializer: string or tf.keras.initializers initializer,
+            defaults to "zeros". The bias initializer for
+            the dense and multiheaded attention layers.
+        name: string, defaults to None. The name of the layer.
+        **kwargs: other keyword arguments.
+
+    Examples:
+
+    ```python
+    inputs = {
+        "tokens": keras.Input(shape=(100,), dtype="int32"),
+        "mask_positions": keras.Input(shape=(20,), dtype="int32"),
+    }
+
+    # Encode the tokens, e.g. with a transformer.
+    encoded_tokens = encoder_model(inputs["tokens"])
+
+    # Predict an output word for each masked input token.
+    outputs = keras_nlp.layers.MLMClassificationHead(
+        vocabulary_size=10000,
+        activation="softmax",
+    )(encoded_tokens, inputs["mask_indices"])
+
+    # Build a model
+    model = keras.Model(inputs, outpus)
+    model.compile(optimizer="adam", loss="sparse_categorical_crossentropy")
```

Comment:
Can we show how a real data would be like? The actual training data has another field `mask_ids`, which is the label input to `model.fit`, representing the original tokens at positions getting masked out. Additionally, `mask_weights` would tell `model.fit()` mask padding information.

Reply:
I switched this to real data and a manually computed loss, basically made this fully runnable in a colab.

I don't think we can show a full end to end training here, that is what our guides are for. We can link the guide from this layer when it is up.